libraries/shared/logs/delegator.go

```diff
@@ -0,0 +1,85 @@
+// VulcanizeDB
+// Copyright © 2019 Vulcanize
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package logs
+
+import (
+	"errors"
+	"github.com/sirupsen/logrus"
+	"github.com/vulcanize/vulcanizedb/libraries/shared/chunker"
+	"github.com/vulcanize/vulcanizedb/libraries/shared/transformer"
+	"github.com/vulcanize/vulcanizedb/pkg/core"
+	"github.com/vulcanize/vulcanizedb/pkg/datastore"
+)
+
+var ErrNoTransformers = errors.New("no event transformers configured in the log delegator")
+
+type ILogDelegator interface {
+	AddTransformer(t transformer.EventTransformer)
+	DelegateLogs(errs chan error, logsFound chan bool)
+}
+
+type LogDelegator struct {
+	Chunker       chunker.Chunker
+	LogRepository datastore.HeaderSyncLogRepository
+	Transformers  []transformer.EventTransformer
+}
+
+func (delegator *LogDelegator) AddTransformer(t transformer.EventTransformer) {
+	delegator.Transformers = append(delegator.Transformers, t)
+	delegator.Chunker.AddConfig(t.GetConfig())
+}
+
+func (delegator *LogDelegator) DelegateLogs(errs chan error, logsFound chan bool) {
+	if len(delegator.Transformers) < 1 {
+		errs <- ErrNoTransformers
+		return
+	}
+
+	persistedLogs, fetchErr := delegator.LogRepository.GetUntransformedHeaderSyncLogs()
+	if fetchErr != nil {
+		logrus.Errorf("error loading logs from db: %s", fetchErr.Error())
+		errs <- fetchErr
+		return
+	}
+
+	if len(persistedLogs) < 1 {
+		logsFound <- false
+	}
+
+	transformErr := delegator.delegateLogs(persistedLogs)
```

Comment:
🤔 I'm good with that - you're basically just saying we don't need to `return` after `errs <- transformErr`, right?

As an aside, I think the error hitting the channel might still bubble up and kill the process, but we could adjust that

Reply:
I'm not saying that would necessarily work ootb, but since transformers themselves set the logs as `transformed`, it seems we have some more freedom to let them fail individually while doing computation on the rest of the set. 

What I am suggesting is passing the error channel as an argument to `delegateLogs`, and letting the individual transformers send their errors there, and `delegateLogs` would continue to execute the rest of the transformers. 

This _does_ move us into `eventual consistency` territory (since we're counting on that missing event to show up later, which it should if it's a temporary error, and it's log is not marked as `transformed`), where we may not want to go. Open question :)